Add document leak tests for Geolocation callbacks

rreno opened the pull request:
#### 968a81c0077cca90309f9aee9bab959597bd4f90
<pre>
Add document leak tests for Geolocation callbacks
<a href="https://bugs.webkit.org/show_bug.cgi?id=276191">https://bugs.webkit.org/show_bug.cgi?id=276191</a>
<a href="https://rdar.apple.com/131057488">rdar://131057488</a>

Reviewed by Ben Nham.

Add tests for navigator.geolocation callbacks to ensure they do not leak
Documents when capturing the document object.

Since the GC is conservative the test robustness is improved by creating
20 iframes and checking that at least one of them does not cause leaks.
It&apos;s possible that the GC sees something pointer-like to the Document
object and therefore keeps it alive. By spamming iframes we should be
able to convince it to collect at least one document, assuming it isn&apos;t
leaked via a reference cycle.

* LayoutTests/http/tests/geolocation/geolocation-get-current-position-does-not-leak.https-expected.txt: Added.
* LayoutTests/http/tests/geolocation/geolocation-get-current-position-does-not-leak.https.html: Added.
* LayoutTests/http/tests/geolocation/geolocation-watch-position-does-not-leak.https-expected.txt: Added.
* LayoutTests/http/tests/geolocation/geolocation-watch-position-does-not-leak.https.html: Added.
* LayoutTests/http/tests/geolocation/resources/geolocation-get-position-callback.html: Added.
* LayoutTests/http/tests/geolocation/resources/geolocation-watch-position-callback.html: Added.

Canonical link: <a href="https://commits.webkit.org/280650@main">https://commits.webkit.org/280650@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e547e71cc3daba6f2e97ba2e3937eaf80e9606a4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57212 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36540 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9687 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60833 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7655 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44164 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7845 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/46321 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5387 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59242 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34287 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49401 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27181 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31069 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6717 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6660 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53036 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/6987 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62513 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1125 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/7078 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/53582 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1130 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49440 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/53654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12643 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/946 "Passed tests") | | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/32369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33454 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/34539 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33200 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham reviewed (approved):
This creates 20 iframes and passes the test if the previous document in the iframe is collected after a navigation to about:blank. It also only fails the test if all 20 of the previous docs in the iframes leaked.

I asked Ryan about this and he said that this is because GC is pretty conservative so just checking to see if a single frame's document is collected could lead to flaky tests. Just writing down so I can remember this in the future.

r+, might be good to explain the rationale behind why the pass/fail is implemented this way, or maybe it's not necessary if there are other existing GC tests that work this way.

rreno committed:
Add document leak tests for Geolocation callbacks
https://bugs.webkit.org/show_bug.cgi?id=276191
rdar://131057488

Reviewed by Ben Nham.

Add tests for navigator.geolocation callbacks to ensure they do not leak
Documents when capturing the document object.

Since the GC is conservative the test robustness is improved by creating
20 iframes and checking that at least one of them does not cause leaks.
It's possible that the GC sees something pointer-like to the Document
object and therefore keeps it alive. By spamming iframes we should be
able to convince it to collect at least one document, assuming it isn't
leaked via a reference cycle.

* LayoutTests/http/tests/geolocation/geolocation-get-current-position-does-not-leak.https-expected.txt: Added.
* LayoutTests/http/tests/geolocation/geolocation-get-current-position-does-not-leak.https.html: Added.
* LayoutTests/http/tests/geolocation/geolocation-watch-position-does-not-leak.https-expected.txt: Added.
* LayoutTests/http/tests/geolocation/geolocation-watch-position-does-not-leak.https.html: Added.
* LayoutTests/http/tests/geolocation/resources/geolocation-get-position-callback.html: Added.
* LayoutTests/http/tests/geolocation/resources/geolocation-watch-position-callback.html: Added.

Canonical link: https://commits.webkit.org/280650@main
